### DISCUSSION TITLE
Rich test runner status prompt

### DISCUSSION THREAD

**Discussion by bartekpacia** [leancodepl/patrol#528] [Ideas]
### Intro

Tools like [Maestro](https://maestro.mobile.dev/) and [Fluttium](https://fluttium.dev/) print test status in the form of easy to comprehend, animated test runner prompt. Go to their websites to see more.

I think Patrol's users would benefit from having a similar task runner.

This is not a priority right now, but I want to start some discussion on this topic to get some feedback.

### Idea

The sample Patrol test

```dart
// integration_test/sign_up_test.dart

patrolTest('signs up', ($) async {
  await $.pumpWidgetAndSettle(AwesomeApp());

  await $(#emailTextField).enterText('bartek@awesome.com');
  await $(#nameTextField).enterText('Bartek');
  await $(#passwordTextField).enterText('ny4ncat');
  await $(#termsCheckbox).tap();
  await $(#signUpButton).tap();
  expect($('Welcome, Bartek!'), findsOneWidget);
});
```

would result in this test runner prompt:

t=0 (start)

```
 ║
 ║  > sign_up_test.dart
 ║
 ║    > signs up
 ║
 ║      ✅ Pump widget and settle
 ║
```

t = 1 (somewhere in the middle)

```
 ║
 ║  > sign_up_test.dart
 ║
 ║    > signs up
 ║
 ║      ✅ Pump widget and settle
 ║      ✅ Enter text "bartek@awesome.com" into widget with key "emailTextField"
 ║      ⌛ Enter text "Bartek" into widget with key "nameTextField"
 ║
```

t=2 (end)

```
 ║
 ║  > sign_up_test.dart
 ║
 ║    > signs up
 ║
 ║      ✅ Pump widget and settle
 ║      ✅ Enter text "bartek@awesome.com" into widget with key "emailTextField"
 ║      ✅ Enter text "Bartek" into widget with key "nameTextField"
 ║      ✅ Enter text "ny4ncat" into widget with key "passwordTextField"
 ║      ✅ Tap on widget with key "termsCheckbox"
 ║      ✅ Tap on widget with key "signUpButton"
 ║      ✅ Expect that one widget with text "Welcome, Bartek!" exists
 ║
```

The above is a pretty blunt copy from Maestro, but I don't think there's lot to be done better.

### Implementation

For now, I just want to shed some light on a single problem: `Finder.toString()` prints what the finder _found_, not what it _represents_. See an example below:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (tester) async {
    await tester.pumpWidget(
      const MaterialApp(
        home: Scaffold(
          body: Center(child: Text('Hello world')),
        ),
      ),
    );

    final finder = find.text('Hello world');
    print(finder);
  });
}
```

Let's see what `Finder.toString()` prints:

```
$ flutter test test/widget_test.dart
00:06 +0: Counter increments smoke test
exactly one widget with text "Hello world" (ignoring offstage widgets): Text("Hello world", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery])
00:06 +1: All tests passed!
```

Problems with it:
- The verbose output is OK for widget tests that developers write, but it is of
  no use to testers.
- The `toString()` actually evaluates the finder and prints what it found in the
  widget tree, instead of printing its representation. 

  In other words, to have nice output, we need a way to differentiate between
  _finder_ and _finder result_.
  
  I don't like how Flutter doesn't make this distinction. I made two examples to
  demonstrate what I mean:

  <details>
    <summary>Simple finder example</summary>

    ```dart
    import 'package:flutter/material.dart';
    import 'package:flutter_test/flutter_test.dart';

    void main() {
      testWidgets('Counter increments smoke test', (tester) async {
        await tester.pumpWidget(
          MaterialApp(
            home: Scaffold(
              body: Center(
                child: Column(
                  children: const [
                    Text('Hello world'),
                    Text('Hello world'),
                  ],
                ),
              ),
            ),
          ),
        );

        final finder = find.text('Hello world');
        print(finder);
      });
    }
    ```

    Current output:

    ```
    $ flutter test test/widget_test.dart
    00:02 +0: Counter increments smoke test
    2 widgets with text "Hello world" (ignoring offstage widgets): [Text("Hello world", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery]), Text("Hello world", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery])]
    00:02 +1: All tests passed!
    ```

    Output I'd like:

    ```
    $ flutter test test/widget_test.dart
    00:02 +0: Counter increments smoke test
    widgets with text "Hello world"
    00:02 +1: All tests passed!
    ```
  </details>

  <details>
    <summary>Nested finder example</summary>

    ```dart
    import 'package:flutter/material.dart';
    import 'package:flutter_test/flutter_test.dart';

    void main() {
      testWidgets('Counter increments smoke test', (tester) async {
        await tester.pumpWidget(
          MaterialApp(
            home: Scaffold(
              body: Center(
                child: Column(
                  children: const [
                    Text('Hello world'),
                    Text('Hello world'),
                  ],
                ),
              ),
            ),
          ),
        );

        final finder = find.descendant(
          of: find.byType(Column),
          matching: find.text('Hello world'),
        );

        print(finder);
      });
    }
    ```

    Current output:

    ```
    $ flutter test test/widget_test.dart
    00:02 +0: Counter increments smoke test
    2 widgets with text "Hello world" that has ancestor(s) with type "Column" (ignoring offstage widgets): [Text("Hello world", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery]), Text("Hello world", dependencies: [DefaultSelectionStyle, DefaultTextStyle, MediaQuery])]
    00:02 +1: All tests passed!
    ```

    Output I'd expect:

    ```
    $ flutter test test/widget_test.dart
    00:02 +0: Counter increments smoke test
    2 widgets with text "Hello world" that has ancestor(s) with type "Column"
    00:02 +1: All tests passed!
    ```
  <details>

  😕

  We could create some extension method `Finder.repr()` which would return the
  "Output I'd like" in the above examples, but it's not as simple as that. The
  `Finder` class has [many subclasses, some of them being
  private](https://github.com/flutter/flutter/blob/6928314d50/packages/flutter_test/lib/src/finders.dart#L590-L651).

### Caveats

- We will only be able to show the past actions and the current one. We don't know what happens in the future. Maestro and Fluttium know that because they interpret their test files. We just execute code.

- This might overlap with our early plans to create a task runner GUI (either as a separate desktop Flutter app, similar to `convenient_test_manager`, or as a VS Code extension).

- Animated task runner prompt could cause problems in environments where the output is not a tty (see #498 and [this](https://github.com/leancodepl/patrol/pull/521#issuecomment-1291242477) to better understand what I mean). We will

**Comment by bartekpacia** [comment]
For now: just print "x test passed, y test failed" at the end of the test.

**Comment by bartekpacia** [comment]
Cross-referencing #561

**Comment by jBorkowska** [comment]
In my opinion it's a bit too much, especially for now. I don't see need for this and it looks complicated. What I need is: if test passed or failed, and (only if failed) why it failed (can be exception as is or just an instruction that couldn't be completed). A list of tests with status and reason of failure would be enough for now.

**Comment by bartekpacia** [comment]
Cross-referencing #602.

**Comment by bartekpacia** [comment]
Cross-referencing https://github.com/flutter/flutter/issues/115874.

**Comment by bartekpacia** [comment]
Live-streaming of native tests lifecycle changes (started, failed, finished) could be implemented with:
- [`InstrumentationRunListener`](https://androidx.tech/artifacts/test/runner/1.6.0-alpha03-source/androidx/test/internal/runner/listener/InstrumentationRunListener.java.html). See [android-xml-run-listener](https://github.com/schroepf/TestLab/tree/master/android/android-xml-run-listener) for an example.
- [`XCTestObservation`](https://developer.apple.com/documentation/xctest/xctestobservation?language=objc) on iOS

Output I have in mind (test name format taken from https://github.com/leancodepl/patrol/issues/1619#issuecomment-1677359318):

```
$ patrol test
...
• Started test: examples__some_example__HomeScreenWhenNotSignedInShowsSignInDialog
✓ Failed test: examples__some_example__HomeScreenWhenNotSignedInShowsSignInDialog
• Started test: examples__some_example__HomeScreenWhenSignedInShowsNewsletterDialog
✓ Succeeded test: examples__some_example__HomeScreenWhenSignedInDoesANiceThing
```

This would be only a development time feature. Patrol CLI would connect to the running app using the existing Dart VM service connection (see #593 and #630 for an example).

**Comment by benjamin-inspectly** [comment]
I know you, @jBorkowska, don't think much of this idea, but I must say it kinda fixes all my current issues with patrol :)
Apparently dart has a bug that "hides" some of the stacktrace as I understand it which gives me the issue I spoke to @bartekpacia  about here: #1615 . It doesn't seem like dart has any plans on fixing that because of a async-issue or whatever... :(

It's not always an option for me to use the adb logcat - especially not if the test has been run on a CI/CD environment - but if I could backtrace the events that patrol triggers on my app it would fix everything.
It doesn't have to be visually pleasing and don't even have to be written to the logs while the test runs, but if I could end up with barteks suggestion alongside the "x-number-of-tests-failed" in the console, I would marry you xD :D

Similar to this one:
![image](https://github.com/leancodepl/patrol/assets/114057836/b572c88f-caf7-4954-87c8-4588bc408e3c)
